### ISSUE TITLE
Morris - ATP for CPU's

### ISSUE THREAD

**Issue by i-rabot** [kbower/tickettest1#2507]
_Imported from trac ticket 2507.
Created by **dwarren**
Opened in trac: Sep 25, 2013, 11:44:53 AM
Last modified in trac: Oct 14, 2015, 11:47:02 AM_
***
_by dwarren:_
_original request:_ Show available inventory based on the Priority fill 

_update:_ At a high level,  ATP is looking at available inventory across the board regardless of where it is located in the DC.  Showing false “available today” for Pickup orders in particular.  Within our fill logic we have pickup orders not reserving from locations that are out of the original manufacture packaging and should only fill if it is a delivery order where we remove items from the box and prep for delivery.

**Comment by i-rabot** [comment]
_**kbower** trac comment on Sep 25, 2013, 12:21:38 PM:_

_from morris:_
Can we just have Pickup Orders ignore locations at 01 in the range of 510000 through 589999 and not report availability of “Today” if in these locations?

**Comment by i-rabot** [comment]
_**kbower** trac comment on Sep 25, 2013, 5:37:53 PM:_

This is far from straightforward.  When lining up orderlines against available inventory, we have to segregate "available inventory" now, something probably not doable by rules.

**Comment by i-rabot** [comment]
_**kbower** trac comment on Sep 26, 2013, 10:22:01 AM:_

Mark,
    I've been looking into this request:

   _Can we just have Pickup Orders ignore locations at 01 in the range
   of 510000 through 589999 and not report availability of "Today" if
   in these locations_

   _  3.Show available inventory based on the Priority fill
   differentiating between Delivery and Pickup reservation. At a high
   level,  ATP is looking at available inventory across the board
   regardless of where it is located in the DC.  Showing false
   "available today" for Pickup orders in particular.  Within our fill
   logic we have pickup orders not reserving from locations that are
   out of the original manufacture packaging and should only fill if it
   is a delivery order where we remove items from the box and prep for
   delivery._
   


This is no trivial matter because, as you know, atp date was a simplistic design in order to quickly calculate a good guess, but a guess nonetheless.  It avoids database hits and intensive calculations because it needs to be recalculated every time a qty is changed, a line added and _even every time an order is loaded._

We may not we able to support this change depending on how much of a re-write of the atp logic is involved.  We also might not _want_ to because it means we need to start fetching more information (that is currently largely cached and so currently doesn't require a database lookup) to figure out which inventory is in that range of locations.  (Currently, it has the cached sum of all available inventory by site.)

This is not a trivial matter of "just ignore that range for pickups" and here is why: since the normal stock can be claimed by both pickups or deliveries, Loft still has to go through all orders and "soft" allocate against the deliveries that might be before the current pickup.  But now it is complicated because it needs to pull from two different (overlapping) pools of inventory based on whether the outstanding unfilled order lines are pickups or deliveries.  Again, this is no minor change to the current simple logic.

I need to know some more information:

1. Do deliveries pull first from that range of locations before going
   to normal stock?
2. Are there some basic assumptions we can make to drastically reduce
   the complexities involved here?  For example, can we just assume
   that there are no outstanding unfilled order lines if available
   inventory exists in the warehouse?  (The rationale being, these
   would have already filled.  If we make that assumption, then we can
   get closer.)
3. Can we assume they won't be asking for the same SKU on one ticket as
   a pickup and other ticket as delivery?
4. Any other simplifying assumptions you can think of to make this
   manageable to Loft?

Thanks,
Kent

**Comment by i-rabot** [comment]
_**kbower** trac comment on Oct 8, 2013, 10:19:38 AM:_

waiting on Morris to answer

**Comment by i-rabot** [comment]
_**kbower** trac status change on Oct 8, 2013, 10:19:38 AM:_ **waiting**

**Comment by i-rabot** [comment]
_**kbower** trac comment on Jan 23, 2014, 9:20:51 AM:_

(Changeset [[4975]](../commit/e277128eaa44a97de0a2468d5ee478e915e27815))
* re #2507, some earlier atp refactoring, but shelving #2507 for now, so committing this much

**Comment by i-rabot** [comment]
_**kbower** trac comment on Oct 14, 2015, 11:47:01 AM:_

Closed with #571

**Comment by i-rabot** [comment]
_**kbower** trac status change on Oct 14, 2015, 11:47:01 AM:_ **closed**

**Comment by i-rabot** [comment]
_**kbower** trac comment on Oct 14, 2015, 11:47:02 AM:_

Closed as duplicate of #571.
